File: file/go/filechain.go
package main

import (
	"fmt"
	"os"
	"path/filepath"
	"time"

	"./blake2b"
)

func main() {
	starttime := time.Now()
	println("\n\t\t\t!!!HashFileChain!!!\n")
	var optslct string
	var pathslct string
	args := os.Args
	switch len(args) { //args check
	case 1:
		helpcheck("<Args check> No option! [FileChain Program] [Option] [Just one folder Path to hash]")
	case 2:
		if args[1] == "help" {
			println("help)\n[FileChain Program] [Option] [Just one folder Path to hash]")
			println("[Option]\n\tnew: 현제 폴더 새로운 해시 생성\n\tupdate: 해시 업데이트")
			println("\tverify: 검증\n\thelp: 사용법\n")
			return
		} else if args[1] == "new" || args[1] == "update" || args[1] == "verify" {
			println("<args check> More Path [FileChain Program] [Option] [Just one folder Path to hash]\n")
			return
		} else {
			helpcheck("<args check> No option! [FileChain Program] [Option] [Just one folder Path to hash]")
		}
	case 3:
		optslct = args[1]
		pathslct = args[2]
		switch optslct { //option check
		case "new":
		case "update":
		case "verify":
		case "help":
			helpcheck("help doesn't need a path.")
		default:
			helpcheck("Plz check option")
		}
		filepath, err := os.Open(pathslct) //path check
		defer filepath.Close()
		errcheck(err, "<path check> Can't open folder")
		fpath, err := filepath.Stat()
		errcheck(err, "<path check> Can't read folder stat")
		if fpath.IsDir() {
			fmt.Printf("OK, Selected! [Option]->%s [Path]->%s\n", optslct, pathslct)
		} else {
			helpcheck("<path check> This is not folder!\n")
		}
	default:
		helpcheck("<args check> Over Args [FileChain Program] [Option] [Just one folder Path to hash]")
	}

	switch optslct { //run
	case "new":
		fc := fchainew(pathslct)
		filechainwrite(pathslct, fc)
	case "update":
		fc := fchainud(pathslct)
		filechainwrite(pathslct, fc)
	case "verify":
		fchainvf(pathslct)
	}
	endtime := time.Since(starttime)
	fmt.Printf("\n실행시간: %s\n", endtime)
}

func fchainvf(path string) {
	var dfstr string
	var dfstrlen int
	var dfhash [64]byte
	check := 0
	var dfes []string
	alldf := allDFfind(path) //폴더안 모든 폴더와 파일경로 검색
	alldflen := len(alldf)
	finalhash := make([]byte, 128) //폴더안 .filechain 값 저장
	for i := 0; i < alldflen; i++ {
		dirfiles, err := os.Open(alldf[i])
		defer dirfiles.Close()
		errcheck(err, "<fchainvf> Can't open file or folder")
		dfstat, err := dirfiles.Stat()
		errcheck(err, "<fchainvf> Can't open file or folder stat")
		dfstr = dfstat.Name()
		dfstrlen = len(dfstr)
		if dfstrlen > 10 && dfstr[dfstrlen-10:] == ".filechain" {
			check++
			_, err := dirfiles.Read(finalhash)
			errcheck(err, "<fchainvf> Can't read *.filechain file")
			continue
		}
		if !dfstat.IsDir() {
			dfstr += fmt.Sprint(dfstat.Size())
			dfstr += fmt.Sprint(dfstat.ModTime().Format("2006-01-02 15:04:05"))
		}
		dfes = append(dfes, dfstr)
	}
	if check > 1 {
		panic("<fchainvf> Why are there more than 2 *.filechain?")
	} else if check == 1 {
		dfeslen := len(dfes)
		for i := 0; i < dfeslen; i++ {
			dfhash = blake2b.Sum512([]byte(dfes[i]))
			for p := 0; p < 64; p++ { //xor로 이전파일체인값 검색 가능
				finalhash[p] ^= dfhash[p]
			}
		}
		truecheck := true
		for t := 0; t < 64; t++ {
			if finalhash[t] != finalhash[t+64] {
				truecheck = false
				println("[verify] -> Flase")
				break
			}
		}
		if truecheck {
			println("[verify] -> True")
		}
	} else {
		panic("<fchainvf> Where is *.filechain?")
	}
}

func fchainud(path string) []byte {
	var dfstr string
	var dfstrlen int
	var dfhash [64]byte
	check := 0
	var dfes []string
	alldf := allDFfind(path) //폴더안 모든 폴더와 파일경로 검색
	alldflen := len(alldf)
	finalhash := make([]byte, 128) //현제 .filechain 값 저장
	for i := 0; i < alldflen; i++ {
		dirfiles, err := os.Open(alldf[i])
		defer dirfiles.Close()
		errcheck(err, "<fchainud> Can't open file or folder")
		dfstat, err := dirfiles.Stat()
		errcheck(err, "<Fchainud> Can't open file or folder stat")
		dfstr = dfstat.Name()
		dfstrlen = len(dfstr)
		if dfstrlen > 10 && dfstr[dfstrlen-10:] == ".filechain" {
			check++
			_, err := dirfiles.Read(finalhash)
			errcheck(err, "<fchainud> Can't read *.filechain file")
			continue
		}
		if !dfstat.IsDir() {
			dfstr += fmt.Sprint(dfstat.Size())
			dfstr += fmt.Sprint(dfstat.ModTime().Format("2006-01-02 15:04:05"))
		}
		dfes = append(dfes, dfstr)
	}
	if check > 1 {
		panic("<Fchainud> Why are there more than 2 *.filechain?")
	} else if check == 1 {
		for m := 0; m < 64; m++ { //현제해시를 이전해시에 덮어쓰기
			finalhash[m+64] = finalhash[m]
		}
		dfeslen := len(dfes)
		for i := 0; i < dfeslen; i++ {
			dfhash = blake2b.Sum512([]byte(dfes[i]))
			for p := 0; p < 64; p++ { //xor로 이전파일체인값 검색 가능
				finalhash[p] ^= dfhash[p]
			}
		}
	} else {
		panic("<Fchainud> Where is *.filechain?")
	}
	return finalhash
}

func fchainew(path string) []byte {
	var dfstr string
	var dfstrlen int
	var dfhash [64]byte
	alldf := allDFfind(path) //폴더안 모든 폴더와 파일경로 검색
	alldflen := len(alldf)
	finalhash := make([]byte, 128) //64byte길이의 초기값 2번 반복해서 입력
	for i := 0; i < alldflen; i++ {
		dirfiles, err := os.Open(alldf[i])
		defer dirfiles.Close()
		errcheck(err, "<fchainew> Can't open file or folder")
		dfstat, err := dirfiles.Stat()
		errcheck(err, "<fchainew> Can't open file or folder stat")
		dfstr = dfstat.Name()
		dfstrlen = len(dfstr)
		if dfstrlen > 10 && dfstr[dfstrlen-10:] == ".filechain" {
			panic("<fchainew> Remove *.filechain file plz")
		}
		if !dfstat.IsDir() {
			dfstr += fmt.Sprint(dfstat.Size())
			dfstr += fmt.Sprint(dfstat.ModTime().Format("2006-01-02 15:04:05"))
		}
		dfhash = blake2b.Sum512([]byte(dfstr))
		for p := 0; p < 64; p++ { //xor로 이전파일체인값 검색 가능
			finalhash[p] ^= dfhash[p]
		}
	}
	return finalhash
}

func filechainwrite(path string, hashvalue []byte) {
	nf, err := os.Create(path + "\\a.filechain") //대상폴더안 파일체인명
	defer nf.Close()
	errcheck(err, "<filechainwrite> Can not create file")
	_, err = nf.Write(hashvalue)
	errcheck(err, "<filechainwrite> Can not write file")
}

func allDFfind(dirlocation string) []string {
	dfinfo := make([]string, 0)
	filepath.Walk(dirlocation, func(path string, info os.FileInfo, err error) error {
		dfinfo = append(dfinfo, path)
		return nil
	})
	return dfinfo
}

func errcheck(e error, s string) {
	if e != nil {
		println(s, "\n")
		os.Exit(1)
	}
}

func helpcheck(s string) {
	println(s, "\nYou can [FileChain Program] [help]\n")
	os.Exit(1)
}

File: file/python/filechain.py
import os
import sys
import timeit
import json
import random
from hashlib import sha3_512
from hashlib import blake2b
from datetime import datetime
import traceback

def help_start():
    print("""
\n\t\t!!!Filechain!!!\n
[OPTION]
    init - make initialization hash.
    set - set file path, added and ignored.
    create - create filechain.
    update - filechain update.
    verify - filechain verify.

<START>""")

    if len(sys.argv)==1:
        raise Exception("Need option.")
    elif len(sys.argv)==2:
        opt = ["init", "set", "create", "update", "verify"]
        if sys.argv[1] in opt:
            print("Args OK.\n")
        else:
            raise Exception("No option.")
    else:
        raise Exception("Args failed.")
    
    return sys.argv[1]

def jsonprint(j):
    print('{')
    for i in j:
        print('  ',i,': ',j[i])
    print('}')

def init():
    init_json = {
        "programname": "",
        "version": "",
        "targetpath": "",
        "savepath": "",
        "initialvalue": "",
        "addfile": [],
        "ignorefile": []
    }
    init_json["programname"] = input("<init - start>\nprogram name: ")
    init_json["version"] = input("program version: ")
    init_json["targetpath"] = input("program target path(only directory): ")
    init_json["savepath"] = input("init json and filechain file path(only directory): ")

    with open(init_json["savepath"]+"\\FilechainInit.json", "w") as f:
        json.dump(init_json, f, indent=2)
    with open(init_json["savepath"]+"\\FilechainAdd.txt", "w") as f:
        pass
    with open(init_json["savepath"]+"\\FilechainIgnore.txt", "w") as f:
        pass
    
    jsonprint(init_json)
    print('<init - success>')

def set():
    json_path = input("<set - start>\nFilechainInit.json path(file): ")
    if json_path.split('\\')[-1] != "FilechainInit.json":
        raise Exception("This file is not FilechainInit.json!")

    init_hash_value = input("Initional hash value(string): ")
    if not init_hash_value:
        init_hash_value = str(random.random())

    with open(json_path, "r") as f:
        json_data = json.load(f)
        json_data["initialvalue"] = init_hash_value
        with open(json_data["savepath"]+"\\FilechainAdd.txt", "r") as adf:
            json_data["addfile"] = list(map(str.strip, adf.readlines()))
        with open(json_data["savepath"]+"\\FilechainIgnore.txt", "r") as igf:
            json_data["ignorefile"] = list(map(str.strip, igf.readlines()))
        json_data["ignorefile"].append(json_data["targetpath"]+f"\\{json_data['programname']}.filechain")#ignore filechain file
    with open(json_path, "w") as f:
        json.dump(json_data, f, indent=2)
    
    jsonprint(json_data)
    print('<set - success>')

def findDaF(dirLocation):#path list return
    dfInfo = []
    dfAppd = dfInfo.append#speed up
    for root, _, files in os.walk(dirLocation):
        for s in files:
            dfAppd(root+'\\'+s)
        dfAppd(root)
    return dfInfo

def create():
    json_path = input("<create - start>\nFilechainInit.json path(file): ")
    if json_path.split('\\')[-1] != "FilechainInit.json":
        raise Exception("This file is not FilechainInit.json!")

    with open(json_path, "r") as f:
        json_data = json.load(f)
        finalHash = sha3_512(json_data["initialvalue"].encode('utf-8')).digest()#Initional hash value
        finalHash += finalHash#make lash hash
        dfList = findDaF(json_data["targetpath"])#make target list, start
        addFiles = json_data["addfile"]
        if addFiles:
            for addFile in addFiles:
                if os.path.isfile(addFile):
                    dfList += addFile
                else:
                    dfList += findDaF(addFile)
        dfList = list(dict.fromkeys(dfList))#deduplication
        ignoreFiles = json_data["ignorefile"]
        if ignoreFiles:
            for ignoreFile in ignoreFiles:
                if os.path.isfile(ignoreFile):
                    dfList.remove(ignoreFile)
                else:
                    for tmp in findDaF(ignoreFile):
                        dfList.remove(tmp)#make target list, end
        
        for x in dfList:#make hash
            if os.path.isfile(x):
                dfStr=f"{os.path.basename(x)}{os.path.getsize(x)}{datetime.fromtimestamp(os.path.getmtime(x)).isoformat(sep=' ', timespec='seconds')}"
            else:
                dfStr=f"{os.path.basename(x)}"
            hash = blake2b(dfStr.encode('utf-8'))
            hash = hash.digest()+bytes(0b1000000)
            finalHash = bytes([a^b for a,b in zip(finalHash, hash)])
        
        with open(json_data["targetpath"]+f"\\{json_data['programname']}.filechain", "wb") as fc:
            fc.write(finalHash)
    print(finalHash)    
    print('<create - success>')

def update():
    json_path = input("<update - start>\nFilechainInit.json path(file): ")
    if json_path.split('\\')[-1] != "FilechainInit.json":
        raise Exception("This file is not FilechainInit.json!")
    
    with open(json_path, "r") as f:
        json_data = json.load(f)
        dfList = findDaF(json_data["targetpath"])#make target list, start
        addFiles = json_data["addfile"]
        if addFiles:
            for addFile in addFiles:
                if os.path.isfile(addFile):
                    dfList += addFile
                else:
                    dfList += findDaF(addFile)
        dfList = list(dict.fromkeys(dfList))#deduplication
        ignoreFiles = json_data["ignorefile"]
        if ignoreFiles:
            for ignoreFile in ignoreFiles:
                if os.path.isfile(ignoreFile):
                    dfList.remove(ignoreFile)
                else:
                    for tmp in findDaF(ignoreFile):
                        dfList.remove(tmp)#make target list, end
        
        with open(json_data["targetpath"]+f"\\{json_data['programname']}.filechain", "rb+") as fc:
            finalHashOriginal = fc.read()
            finalHash = finalHashOriginal[:64]+finalHashOriginal[:64]#nowhash to lasthash
            for x in dfList:
                if os.path.isfile(x):
                    dfStr=f"{os.path.basename(x)}{os.path.getsize(x)}{datetime.fromtimestamp(os.path.getmtime(x)).isoformat(sep=' ', timespec='seconds')}"
                else:
                    dfStr=f"{os.path.basename(x)}"
                hash = blake2b(dfStr.encode('utf-8'))
                hash = hash.digest()+bytes(0b1000000)
                finalHash = bytes([a^b for a,b in zip(finalHash, hash)])
            if finalHashOriginal == finalHash[64:]+finalHash[:64]:#not change target file
                print('<verify - failure>')
                return False
            else:
                fc.seek(0)
                fc.write(finalHash)
                print('<update - success>')
                return True

def verify():
    json_path = input("<verify - start>\nFilechainInit.json path(file): ")
    if json_path.split('\\')[-1] != "FilechainInit.json":
        raise Exception("This file is not FilechainInit.json!")
    
    verifyCheck = True
    with open(json_path, "r") as f:
        json_data = json.load(f)
        dfList = findDaF(json_data["targetpath"])#make target list, start
        addFiles = json_data["addfile"]
        if addFiles:
            for addFile in addFiles:
                if os.path.isfile(addFile):
                    dfList += addFile
                else:
                    dfList += findDaF(addFile)
        dfList = list(dict.fromkeys(dfList))#deduplication
        ignoreFiles = json_data["ignorefile"]
        if ignoreFiles:
            for ignoreFile in ignoreFiles:
                if os.path.isfile(ignoreFile):
                    dfList.remove(ignoreFile)
                else:
                    for tmp in findDaF(ignoreFile):
                        dfList.remove(tmp)#make target list, end
        
        with open(json_data["targetpath"]+f"\\{json_data['programname']}.filechain", "rb") as fc:
            finalHash = fc.read()
            for x in dfList:
                if os.path.isfile(x):
                    dfStr=f"{os.path.basename(x)}{os.path.getsize(x)}{datetime.fromtimestamp(os.path.getmtime(x)).isoformat(sep=' ', timespec='seconds')}"
                else:
                    dfStr=f"{os.path.basename(x)}"
                hash = blake2b(dfStr.encode('utf-8'))
                hash = hash.digest()+bytes(0b1000000)
                finalHash = bytes([a^b for a,b in zip(finalHash, hash)])
            for p in range(64):#verify check
                if finalHash[p] != finalHash[p+64]:
                    verifyCheck = False
                    break
    if verifyCheck:
        print('<verify - success>')
        return True
    else:
        print('<verify - failure>')
        return False

if __name__=="__main__":
    starttime = timeit.default_timer()

    try:
        argv = help_start()
        if argv == "init":
            init()
        elif argv == "set":
            set()
        elif argv == "create":
            create()
        elif argv == "update":
            update()
        elif argv == "verify":
            verify()
    except Exception as e:
        print('<err>: ', e)
        print(traceback.format_exc())
    finally:
        print("\n<END>\nProgram exit.")

    endtime = timeit.default_timer()
    print(f"Running Time: {endtime-starttime}ms")
